docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,317 @@
+
+### What else can be done
+
+**Omit rewriting .NET binaries**
+
+This means that if _for some reason_ the recompilation is triggered but there is nothing to retypecheck, reuse the binaries. This shouldn't be hard to do (just writing and reading existing files) - but it's unclear how often this happens in real life.
+
+**Omit all the other steps**
+
+If we have universal caching mechanism (see below), we can apply it to other steps (e.g. parsing). Those steps typically don't take much time each on their own, but can somewhat accummulate and maybe even peak in some edge cases. 
+
+**Integrate this into tooling**
+
+We can plug the feature into the tooling. This would lead to time savings in many more developer scenarios. However, this is not trivial and the tooling layer of the feature would need to do extra checks and invalidation calls, e.g. when the tooling and the compiler use different versions of the compiler service.
+
+**Omit recompilation for insignificant file changes**
+
+That is about getting smart after getting first parts of the compilation results. If we know that the code is IL-same, we can skip everything and reuse the binaries. This would be useful in scenarios like adding comments or removing unused opens. However, this might present a **hard** (albeit interesting) challenge - things like figuring out AST diffs seem to be unexplored yet.
+
+## Open questions
+
+### How to cache data?
+
+We can store the data in the intermediate build directory (e.g. `obj` in case of `dotnet build`). The usual way for caching would be to serialize + deserialize information.
+
+However, serialization of the compiler data is non-trivial:
+- we don't have a **full** out-of-the-box solution in the repo (some parts are available, see below)
+- well-known serialization libraries (like `System.Text.Json`) won't work right away, because data is internal and libs can't handle some F# type easily (e.g. DUs)
+- we can't use F#-specific libraries (like [this](https://github.com/fsprojects/FSharp.Json) or [this](https://github.com/Tarmil/FSharp.SystemTextJson)) because they are not "official" enough
+
+It should be possible to hack things around, but the more compiler data we want to serialize, the harder it will be. Hence I see 2 ways to go:
+
+**1) Start with leveraging some existing elements and then see.** 
+
+We have a mechanism for writing signature files. We could force generating signatures for fresh compilation, save them and then "feed" to the recompilation process. This will speed up typechecking, saving a lot of time on type inference. We also seem to have a mechanism for encoding optimization data. We can apply the same idea here - force saving it for the first run and then utilize for reruns. This won't help us with everything but **might be** good enough. Also, the things we'll be left with **might be** easier to hand-serialize.
```

Comment:
This paragraph is too vague, it would be a lot clearer if it used actual terms from the compiler codebase.
- Which types would be missing considering the combined scope of signature+optdata pickler?

Reply:
Remove for now, pickling + unpickling might be enough.